applications/minotari_node/src/stratum/block_template_repository.rs

```diff
@@ -0,0 +1,329 @@
+            .await?;
+
+        total_nonce = &total_nonce +
+            &new_nonce
+                .pub_key
+                .to_public_key()
+                .map_err(|e| anyhow!("Failed to get public key: {}", e))?;
+        total_excess = &total_excess +
+            &coinbase_kernel
+                .excess
+                .to_commitment()
+                .map_err(|e| anyhow!("Failed to get commitment: {}", e))?;
+        let (spending_key_id, nonce) = (wallet_output.commitment_mask_key_id, new_nonce.key_id);
+        let kernel_message = TransactionKernel::build_kernel_signature_message(
+            &TransactionKernelVersion::get_current_version(),
+            coinbase_kernel.fee,
+            coinbase_kernel.lock_height,
+            &coinbase_kernel.features,
+            &None,
+        );
+        let mut kernel_signature = UncompressedSignature::default();
+        kernel_signature = &kernel_signature +
+            &key_manager
+                .get_partial_txo_kernel_signature(
+                    &spending_key_id,
+                    &nonce,
+                    &CompressedPublicKey::new_from_pk(total_nonce.clone()),
+                    &CompressedPublicKey::new_from_pk(total_excess.as_public_key().clone()),
+                    &TransactionKernelVersion::get_current_version(),
+                    &kernel_message,
+                    &coinbase_kernel.features,
+                    TxoStage::Output,
+                )
+                .await?
+                .to_schnorr_signature()
+                .map_err(|e| anyhow!("Failed to get schnorr signature: {}", e))?;
+        let kernel_new = KernelBuilder::new()
+            .with_fee(0.into())
+            .with_features(coinbase_kernel.features)
+            .with_lock_height(coinbase_kernel.lock_height)
+            .with_excess(&CompressedCommitment::from_commitment(total_excess))
+            .with_signature(CompressedSignature::new_from_schnorr(kernel_signature))
+            .build()
+            .unwrap();
+
+        new_template.body.add_kernel(kernel_new);
+        new_template.body.sort();
+
+        let new_block = match handler.get_new_block(new_template).await {
+            Ok(b) => b,
+            Err(e) => {
+                warn!(target: LOG_TARGET, "Failed to get new block: {}", e);
+                return Err(anyhow!("Failed to get new block: {}", e));
+            },
+        };
+        let mining_hash = match new_block.header.pow.pow_algo {
+            PowAlgorithm::Sha3x => new_block.header.mining_hash().to_vec(),
+            PowAlgorithm::RandomXT => new_block.header.mining_hash().to_vec(),
+            PowAlgorithm::RandomXM => new_block.header.merge_mining_hash().to_vec(),
+            PowAlgorithm::Cuckaroo => new_block.header.mining_hash().to_vec(),
+        };
+        let vm_key = *handler
+            .get_header(tari_rx_vm_key_height(new_block.header.height))
+            .await?
+            .unwrap()
```

Comment:
`unwrap` should be handled
```suggestion
            .unwrap()
```